### PR TITLE
SAMZA-2562: [Scala cleanup] Clean up Scala in samza-kv-inmemory and samza-kv-couchbase

### PR DESCRIPTION
**Pull request by cameronlee314**
Issues:
1) `samza-kv-couchbase` uses the Scala Gradle plugin but does not have Scala code
2) `samza-kv-inmemory` is in Scala, but we are no longer using Scala in Samza, so it would be good to move to Java.
3) `samza-kv-inmemory` is missing unit tests

Changes:
1) Rewrite `samza-kv-inmemory`in Java (no functional changes)
2) Remove usage of Scala Gradle plugin for `samza-kv-couchbase` and `samza-kv-inmemory`
3) Add unit tests for `InMemoryKeyValueStore` in `samza-kv-inmemory`.

Tests:
1) Added unit tests for original Scala version of `InMemoryKeyValueStore` and verified that they passed.
2) Tests pass for new Java version of `InMemoryKeyValueStore`.
3) `./gradlew build` (some of the tests in `samza-test` use `InMemoryKeyValueStore`).

API changes: I don't think the `InMemoryKeyValueStore` is part of the Samza API, since `InMemoryKeyValueStorageEngineFactory` is supposed to be used, but the `InMemoryKeyValueStore` constructor did change from having one `KeyValueStoreMetrics` argument with a default value of `new KeyValueStoreMetrics` to having a required `KeyValueStoreMetrics` argument. I believe this only applies to Scala usage of `InMemoryKeyValueStore`, since Java usage requires all arguments (even ones with default values) to be passed anyways.

Upgrade/usage instructions: If using the `InMemoryKeyValueStore` constructor in another Scala class, then add the `KeyValueStoreMetrics` argument. It's unlikely that anyone is depending on this constructor explicitly.

Notes:
While writing the unit tests, I found a couple of bugs: https://issues.apache.org/jira/browse/SAMZA-2563 (InMemoryKeyValueStore double counts deletes in metrics) and https://issues.apache.org/jira/browse/SAMZA-2564 (InMemoryKeyValueStore.snapshot does not satisfy immutability API). I did not include fixes in this PR, since I want to keep the Scala refactoring separate from any functional changes.
This PR is not intended to have any functional changes.

### CODE REVIEW

**Comment by cameronlee314** [comment]
@bkonold Could you please review?

File: samza-kv-inmemory/src/main/java/org/apache/samza/storage/kv/inmemory/InMemoryKeyValueStore.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    // TreeMap's putAll requires a map, so we'd need to iterate over all the entries anyway
+    // to use it, in order to putAll here.  Therefore, just iterate here.
+    for (Entry<byte[], byte[]> next : entries) {
+      put(next.getKey(), next.getValue());
+    }
+  }
+
+  @Override
+  public void delete(byte[] key) {
+    // TODO Bug: This double counts deletes for metrics, because put also counts a delete
```
Comment (bkonold) [inline samza-kv-inmemory/src/main/java/org/apache/samza/storage/kv/inmemory/InMemoryKeyValueStore.java:90]: link the jira?

File: samza-kv-inmemory/src/main/java/org/apache/samza/storage/kv/inmemory/InMemoryKeyValueStore.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  public KeyValueIterator<byte[], byte[]> range(byte[] from, byte[] to) {
+    this.metrics.ranges().inc();
+    Preconditions.checkArgument(from != null, "Null argument 'from' not allowed");
+    Preconditions.checkArgument(to != null, "Null argument 'to' not allowed");
+    return new InMemoryIterator(this.underlying.subMap(from, to).entrySet().iterator(), this.metrics);
+  }
+
+  @Override
+  public KeyValueSnapshot<byte[], byte[]> snapshot(byte[] from, byte[] to) {
+    // TODO: Bug: This does not satisfy the immutability constraint, since the entrySet is backed by the underlying map.
```
Comment (bkonold) [inline samza-kv-inmemory/src/main/java/org/apache/samza/storage/kv/inmemory/InMemoryKeyValueStore.java:105]: link the jira?

File: samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java [hunk 3]
```diff
@@ -19,66 +19,501 @@
...
+
+  @Test
+  public void testPutEmpty() {
+    byte[] emptyValue = new byte[0];
+    this.inMemoryKeyValueStore.put(key(0), emptyValue);
+
+    assertEquals(0, this.inMemoryKeyValueStore.get(key(0)).length);
+    verify(this.putsCounter).inc();
+    verify(this.bytesWrittenCounter).inc(key(0).length);
+  }
```
Comment (bkonold) [inline samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java:146]: is there an edge case this is intended to test that `testPut` doesn't cover?

File: samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java [hunk 4]
```diff
@@ -19,66 +19,501 @@
...
+    }
+    this.inMemoryKeyValueStore.putAll(entries);
+
+    for (int i = 0; i < 10; i++) {
+      assertArrayEquals(value(i), this.inMemoryKeyValueStore.get(key(i)));
+    }
+    // 1 time for initial value to be overridden, 10 times for "regular" puts
+    verify(this.putsCounter, times(11)).inc();
+    verify(this.bytesWrittenCounter).inc(key(0).length + value(1234).length);
+    verify(this.bytesWrittenCounter, times(10)).inc(key(0).length + value(0).length);
```
Comment (bkonold) [inline samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java:190]: doesn't this assume that `key(0)` and `value(0)` lengths are the same as `key(i)` and `value(i)` from the loop? can we re-write this so it more obviously matches the 10 writes that happen in the loop or add a comment indicating this assumption? or perhaps define a constant that represents the length of an integer key / value... just to make this more readable.

same comment applies to other occurrences of loops in your tests

File: samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java [hunk 5]
```diff
@@ -19,66 +19,501 @@
...
+  }
+
+  @Test
+  public void testDelete() {
+    this.inMemoryKeyValueStore.put(key(0), value(0));
+    this.inMemoryKeyValueStore.delete(key(0));
+    assertNull(this.inMemoryKeyValueStore.get(key(0)));
+
+    /*
+     * There is a bug in which deletes are double counted in metrics. This deletesCounter should only be invoked once
```
Comment (bkonold) [inline samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java:229]: ty for noting the bug; will make it easy to fix when tests fail once the bug is fixed

File: samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java (hunk 3 above)
Comment (cameronlee314) [reply samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java:146]: Yes, this tests when the value has zero length. `testPut` is puts a non-empty value.

Comment (bkonold) [reply samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java:146]: Sorry should have been more clear. What code path is this testing that testPut isn't? Does the implementation behave differently when the length is 0?

File: samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java (hunk 4 above)
Comment (cameronlee314) [reply samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java:190]: Yep, it does assume that, good point. Thanks for the suggestion. I defined a constant and added some more javadocs to clarify this.

File: samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java (hunk 3 above)
Comment (cameronlee314) [reply samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java]: With the current implementation, it doesn't test a different code path. It just seemed like an edge case that would be good to cover.
